README.TXT
===========
pyEventAnalysis 
===========

Description::

Running the code requires a python file as shown below and a settings file located in the data directory. 



Installation
=========

The code is tested under OS X and Ubuntu. Installation instructions for both platforms are below.

Dependencies
-------------
The following manual dependencies are needed before the code will run. Once they are installed, type ‘sh dependencies/dependencies/build-deps-sh’ to install additional dependencies.

* OS X
	First install Homebrew. Then install python using brew
		brew install python

	Next install the Fortran compiler from <http://hpc.sourceforge.net/>

* Ubuntu
	Under Ubuntu in addition to gfortran, the BLAS and LAPACK libraries are required
apt-get install libblas-dev liblapack-dev gfortran libfreetype6-dev

Mathematica Scripts
-------------

To install the Mathematica analysis scripts type "make math-iface" from the root of the source directory.

Testing the Installation
-------------

Type "make tests" from the root of the source directory to test if the code was installed correctly.


Settings
=========

The code processes events in two stages. The first stage is a simple threshold that identifies events. This stage is controlled by the ‘eventSegment’ section in the settings file. The second stage performs a least squares fit of the model to each identified event and is controlled by the ‘stepResponseAnalysis’ section.

{
        "eventSegment" : {
                "blockSizeSec"          : "0.5",
                "eventPad"              : "50",
                "minEventLength"        : "5",
                "eventThreshold"        : "4.0",
                "driftThreshold"        : "999",
                "maxDriftRate"          : "999",
                "meanOpenCurr"          : "134.0",
                "sdOpenCurr"            : "4.0",
                "slopeOpenCurr"         : "0",
                "writeEventTS"          : "1",
                "parallelProc"          : "0",
                "reserveNCPU"           : "3",
                "plotResults"           : "0"
        },
        "stepResponseAnalysis" : {
                "FitTol"                : "1.e-7",
                "FitIters"              : "10000",
                "BlockRejectRatio"      : "0.9"
        }
}

Running the Code
=========

The code should be run from the root of the source directory. Type python bin/analysis.py. The analysis.py file  is below:
   	import sys
	import os
	import glob
	import time
	import signal
	import resource

	import pyeventanalysis.SingleChannelAnalysis

	import pyeventanalysis.eventSegment as es

	import pyeventanalysis.stepResponseAnalysis as sra 
	import pyeventanalysis.multiStateAnalysis as msa

	from pyeventanalysis.qdfTrajIO import *
	from pyeventanalysis.abfTrajIO import *

	from pyeventanalysis.besselLowpassFilter import *



	pyeventanalysis.SingleChannelAnalysis.SingleChannelAnalysis(
				abfTrajIO(dirname='/Volumes/DATA/PEG1500/PEG29/',filter='*abf', nfiles=1, start=1000), 
				es.eventSegment,
				sra.stepResponseAnalysis
	 		).Run()

CHANGELOG.rst
Change Log
++++++++++++++++++++++++++++++++++++++++++

**v2.4**
- Maintenance release to update dependencies, unit testing frameworks and streamline setuptools installation.
- Resolves `#110 <https://github.com/usnistgov/mosaic/issues/110>`_, `#111 <https://github.com/usnistgov/mosaic/issues/111>`_, `#112 <https://github.com/usnistgov/mosaic/issues/112>`_ and `#116 <https://github.com/usnistgov/mosaic/issues/116>`_, 

**v2.2**

- Updated web UI to allow lowpass filtering (BesselLowpassFilter).
- Allow data source path to be changed from web UI when running in local mode (issue `#109 <https://github.com/usnistgov/mosaic/issues/109>`_).

**v2.1**

- Fixed a bug with dcOffset correction that caused the web UI to not accurately display the trajectory.
- Provide Docker images to allow cross-platfom installs
- Bug fixes

**v2.0**

- Fixed an issue where spurious events were detected that were contained entirely within the open channel current (see `#102 <https://github.com/usnistgov/mosaic/issues/102>`_).
- Changed eval calls to safer `AST <https://docs.python.org/2/library/ast.html>`_-based versions within the core mosaic modules.
- Added detailed Windows build instructions.

**v2.0b1**

- Run MOSAIC using a new web interface.

**v1.3.6**

- Bug Fixes

**v1.3.5**

- Saves raw events in database. Resolves `#93 <https://github.com/usnistgov/mosaic/issues/93>`_.

**v1.3.4**

- Fixed a bug in the default settings that prevented the analysis from starting in some instances.

**v1.3.3**

- Added support for Chimera data files.
- Resolves issues `#92 <https://github.com/usnistgov/mosaic/issues/92>`_ and `#95 <https://github.com/usnistgov/mosaic/issues/95>`_.

**v1.3.2**

- Fixed event partiioning bugs.
- [GUI] UI tweaks.

**v1.3.1**

- Bug fix release.

**v1.3**

- Added a new MOSAIC examples repository that contains plotting scripts and Jupyter notebooks (https://github.com/usnistgov/mosaic-examples)

- Updated Bessel fiter implementation to allow causal and non-causal filters.
- Added the CUSUM+ algorithm (see pull requests #34, #43, #45, and #46)
- Added the ability to read arbitrary binary files.
- Renamed StepResponseAnalysis to ADEPT 2-state.
- Renamed multiStateAnalysis to ADEPT.
- CUSUM+ is now used to generate initial guesses for ADEPT
- Streamlined unit test framework. Added new tests for CUSUM+.
- Updated MOSAIC dependencies to include newer package versions and pandas. Run 'python setup.py mosaic_deps' to update.
- Added a new metadata column (mdStateResTime) that saves the residence time of each state to the database. This affects ADEPT and CUSUM+.
- Processing time per event is now saved to the database.
- Added support for warning status codes that begin with 'w'. These codes preserve metadata in the SQLite output and are used to flag events that need closer inspection.
- Removed mosaicgui from PyPi. 'pip install mosaic-nist' only installs command line modules. 
- Top level ConvertTrajIO supports binary and ASCII files.
- Added Travis continous integration support.
- [GUI] Added a binary update service for Windows and OS X.
- [GUI] Added a table that displays fit parameters in the event viewer window.
- [GUI] Added more descriptive fit errors.
- [GUI] Added CUSUM+ support to MOSAIC GUI.
- [GUI] Export SQLITE database to CSV (#56).
- [GUI] Support for warning codes in fit event viewer.
- [GUI] Performance optimizations.
- [GUI] Fit window in MOSAIC GUI displays idealized pulses overlays.
- [GUI] Added additional analysis statistics.

- [Addons] Added CUSUM+ support to Mathematica addon (PlotEvents in MosaicUtils.m)
- [Addons] Mathematica queries are *optionally* handled through an external Python script to improve performance.
- [Addons] Added an option to limit PlotEvents in Mathematica addon to N events.

- Known Issues: See #8, #10, #40, #63 and #69.

**v1.2**

- Added support for arbitrary binary file formats (#33)
- [GUI] Included binary file support.
- Documentation updates and bug fixes.
- *Known Issues:* See #8 and #10.

**v1.1**

- [Addons] IGOR support.
- PyPi package automatically installs MOSAIC dependencies.
- Miscellaneous bug fixes.
- *Known Issues:* See #8 and #10.

**v1.0**

- Fixed a bug in multistate code that constrained the RC constant resulting in systematic fitting errors (pull request #25).
- Updated multistate to include a separate RC constant for each state, to be consistent with functional form in Balijepalli et al., ACS Nano 2014.
- Misc bug fixes in tsvTrajIO parsing.
- The number of states is saved to the MDIO DB for multistate analysis (issue #26).
- Created a new package on PyPI (mosaic-nist) to allow installation with setuptools.
- [GUI] Updated help link to point to Sphinx documentation on Github.
- *Known Issues:* See #8 and #10 

**v1.0b3.2**

- [GUI] Misc bug fixes
- [Addons] Added code to import MOSAIC output into Matlab (pull requests #18 and #20)
- [Addons] Updated Mathematica addons to automatically decode multi-state data.
- Resolves issues #16 and #22

**v1.0b3.1**

- [GUI] Added multiState support to mosaicgui.
- Analysis information such as alogirthms used, data type, etc. are now stored within a MDIO database.
- [GUI] Autocomplete in mosaicgui only suggests database columns that are valid when used in a query.
- Reorganized Mathematica addon code.


**v1.0b3**

- Fixed a bug that prevented events longer than ~700 data points from being correctly analyzed.
- Fixed a problem that prevented event data from being correctly padded before analysis.
- Resolves #2. TrajIO settings are now read in from the settings file.
- [GUI] Resolves #3. Threshold entry box in GUI becomes nonresponsive when meanOpenCurr is negative.
- [GUI] Resolves #4. Analysis fails when using wavletDenioseFilter from GUI.
- [GUI] Histogram in BlockDepthViewer window can be saved to a CSV file from the File Menu.

download.py
import mosaic
from mosaic.utilities.resource_path import resource_path

def GenerateTitle():
  return """
    <div flex layout="column" tabIndex="-1" role="main" class="md-whiteframe-z3">

      <md-toolbar ng-controller="AppCtrl" layout="row" hide-sm class="md-toolbar-tools md-toolbar-tools-center md-medium-tall md-whiteframe-z2" style="background-color: white;">
        <a href="index.html" style="padding-top: 15px; padding-left: 25px;"><md-icon md-font-set="material-icons md-24">arrow_back</md-icon></img></a>
        <span flex></span>
        <md-button ng-href="{{docURL}}"><h4>Documentation</h4></md-button>
        <md-button ng-href="{{mailingListURL}}"><h4>Mailing List</h4></md-button>
        <md-button ng-href="https://github.com/usnistgov/mosaic"><h4>Develop</h4></md-button>
        <md-button ng-href="https://github.com/usnistgov/mosaic/issues"><h4>Issue Tracker</h4></md-button>
      </md-toolbar>

      <md-toolbar layout="row" hide-gt-sm class="md-whiteframe-z4" layout-align="center center">
        <md-button class="md-icon-button" aria-label="Settings" ng-click="toggleLeft()">
          <md-icon md-font-set="material-icons md-24">menu</md-icon>
        </md-button>

        <a href="index.html" style="text-decoration: none;"><h3><em>MOSAIC</em></h3></a>
        <span flex></span>
      </md-toolbar>

      <md-sidenav class="sidenav md-sidenav-left md-whiteframe-z2" md-component-id="left">
        <md-toolbar class="md-tall" style="background-color: #283593;">
          <span flex></span>
          <h1 class="md-toolbar-tools md-toolbar-tools-bottom">
            <span class="md-flex">MOSAIC</span>
          </h1>
        </md-toolbar>
        <md-content ng-controller="LeftCtrl" class="sidenav">
          <a href="{{docURL}}"><h2>Documentation</h2></a>
          <a href="{{mailingListURL}}"><h2>Mailing List</h2></a>
          <a href="https://github.com/usnistgov/mosaic"><h2>Develop</h2></a>
          <a href="https://github.com/usnistgov/mosaic/issues"><h2>Issue Tracker</h2></a>
        </md-content>
      </md-sidenav>

      <md-content flex id="content">
          <div layout="row" class="download-page" layout-align="center center">
            <div flex="10" flex-sm="20"><img src="assets/img/icon100px.png" width="50%"></img></div>
            <div flex><h2>Downloads</h2></div>
          </div>
  """

def GenerateHeader():
  return """
  <head>
    <title>MOSAIC Downloads</title>
    <link rel="icon" href="assets/img/favicon.ico" type="image/ico"/>

    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
    <meta http-equiv="Content-Language" content="en"/>
    <meta http-equiv="Content-Style-Type" content="text/css"/>
    <meta name="description" content="MOSAIC is a modular single-molecule analysis toolbox to decode multi-state nanopore data."/>
    <meta name="author" content="Arvind Balijepalli"/>
    <meta name="google-site-verification" content="BXETzCWmp_8xdMvDlUidbzJ4zbvyjVnNwiabBPnCvDk"/>
    <meta name="msvalidate.01" content="6486B1AC4BD85BB7125DD2B47CA4E4C5"/>

    <meta name="viewport" content="initial-scale=1, maximum-scale=1, user-scalable=no" />

    <link rel='stylesheet' href='https://fonts.googleapis.com/css?family=Roboto:200,300,400,400italic'>
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    <link rel="stylesheet" href="https://ajax.googleapis.com/ajax/libs/angular_material/1.1.8/angular-material.min.css">

    <link rel="stylesheet" href="assets/app.css"/>

    <!-- JSON-LD markup generated by Google Structured Data Markup Helper. -->
		<script type="application/ld+json">
		{
		  "@context" : "http://schema.org",
		  "@type" : "SoftwareApplication",
		  "name" : "MOSAIC",
		  "image" : "https://pages.nist.gov/mosaic/assets/img/icon.png",
		  "url" : "https://pages.nist.gov/mosaic",
		  "author" : {
		    "@type" : "Person",
		    "name" : "Arvind Balijepalli"
		  },
		  "datePublished" : "2018-08-16",
		  "publisher" : {
		    "@type" : "Organization",
		    "name" : "NIST"
		  },
		  "applicationCategory" : "Single-molecule analysis",
		  "downloadUrl" : "https://pages.nist.gov/mosaic/download.html",
		  "operatingSystem" : "Windows; Mac OS",
		  "screenshot" : "https://pages.nist.gov/mosaic/assets/img/GUI-01.png",
		  "softwareVersion" : "2.0",
  		  "offers": {
		    "@type": "Offer",
		    "availability": "http://schema.org/InStock",
		    "price": "0.00",
		    "priceCurrency": "USD"
		  }
		}
    </script>
		<script async type="text/javascript" id="_fed_an_ua_tag" src="https://dap.digitalgov.gov/Universal-Federated-Analytics-Min.js?agency=DOC&subagency=NIST&pua=UA-66610693-1&yt=true&exts=ppsx,pps,f90,sch,rtf,wrl,txz,m1v,xlsm,msi,xsd,f,tif,eps,mpg,xml,pl,xlt,c"></script>
  </head>
  
  <body ng-app="mosaicApp" layout="row" ng-controller="AppCtrl">
  """
def GenerateTableFromVersion(baseurl, title, version, build, showBinaries=True):
  tabsrc="""
    <md-list>
            <md-subheader class="md-no-sticky"><h3>{0}</h3></md-subheader>
    """.format(title)
  if showBinaries:
    tabsrc+="""
      <md-list-item class="md-1-line">
              <div flex class="md-list-item-text download-links" layout="row">
                <div flex><a href="{0}/mosaic-{1}{2}.dmg"><h4>mosaic-v{1}{2}.dmg</h4></a></div>
                <div flex><h4>Mac OS X, 10.8+</h4></div>
              </div>
            </md-list-item>

            <md-list-item class="md-1-line">
              <div flex class="md-list-item-text download-links" layout="row">
                <div flex><a href="{0}/mosaic-x64-{1}{2}.zip"><h4>mosaic-x64-{1}{2}.zip</h4></a></div>
                <div flex><h4>Windows, 64-bit</h4></div>
              </div>
            </md-list-item>
    """.format(baseurl, version, build)
  tabsrc+="""
      <md-list-item class="md-1-line">
              <div flex class="md-list-item-text download-links" layout="row">
                <div flex><a href="{0}/mosaic-nist-{1}{2}.tar.gz"><h4>mosaic-nist-{1}{2}.tar.gz</h4></a></div>
                <div flex><h4>MOSAIC Source</h4></div>
              </div>
            </md-list-item>""".format(baseurl, version, build)
  if showBinaries:
    tabsrc+="""
            <md-list-item class="md-1-line">
              <div flex class="md-list-item-text download-links" layout="row">
                <div flex><a href="https://github.com/usnistgov/mosaic/releases/download/v1.0/MOSAICReferenceData.zip"><h4>mosaic-reference-data.zip</h4></a></div>
                <div flex><h4>MOSAIC Reference Data</h4></div>
              </div>
            </md-list-item>"""
  tabsrc+="""
      <md-divider ></md-divider>
        </md-list>
  """

  return tabsrc

if __name__ == '__main__':
  with open( resource_path('version-hash'), 'r' ) as f:
      version=f.read().strip()

  url="https://github.com/usnistgov/mosaic/releases/download/v"
  vers=[

    ("Version 2.2", "2.2", "+06ecb7a", True),
    ("Version 2.1", "2.1", "+2f0cf5a", False),
    ("Version 2.0b1", "2.0b1", ".f98ae80", False),
    # ("Version "+str(version), str(version), ".f98ae80", True),
    ("Version 1.3.6", "1.3.6", ".0a2ea45", False),
    ("Version 1.3.5", "1.3.5", ".801e7e5", False),
    ("Version 1.3.4", "1.3.4", ".4932336", False),
    ("Version 1.3.2", "1.3.2", ".d88866c", False),
    ("Version 1.3.1", "1.3.1", ".9977009", False),
    ("Version 1.3", "1.3", ".f0e754c", False),
    ("Version 1.2", "1.2", "", False),
    ("Version 1.1", "1.1", "", False),
    ("Version 1.0", "1.0", "", False)
    ]
  htmlout="""
  <!DOCTYPE html>
  <html lang="en">
  """

  htmlout+=GenerateHeader()
  htmlout+=GenerateTitle()
  for v in vers:
    htmlout+=GenerateTableFromVersion(url+v[1], v[0], v[1], v[2])
  htmlout+="""
    </md-content>
    </div>
 
     <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.6.9/angular.min.js"></script>
    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.6.9/angular-animate.min.js"></script>
    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.6.9/angular-aria.min.js"></script>

    <!-- Angular Material Library -->
    <script src="https://ajax.googleapis.com/ajax/libs/angular_material/1.1.8/angular-material.min.js"></script>

    <script src="mosaic.js"></script>

  </body>
</html>"""

  with open('download.html', 'w') as f:
    f.write(htmlout)

  print("Wrote download.html")

